<a href="https://colab.research.google.com/github/sebekpro/python-krakowiak/blob/main/250-data-science-bootcamp-w-jezyku-python/221_230_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Science Bootcamp

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 221](#1)
* [Ćwiczenie 222](#2)
* [Ćwiczenie 223](#3)
* [Ćwiczenie 224](#4)
* [Ćwiczenie 225](#5)
* [Ćwiczenie 226](#6)
* [Ćwiczenie 227](#7)
* [Ćwiczenie 228](#8)
* [Ćwiczenie 229](#9)
* [Ćwiczenie 230](#10)

### <a name='0'></a>Import biblioteki

In [ ]:
import numpy as np
np.random.seed(20)

np.__version__

'2.0.2'

### <a name='a11'></a> Rozwiązywanie układów równań
Rozważmy układ równań $U$:  
$${\displaystyle \mathrm {U} \colon {\begin{cases}{\begin{matrix}a_{11}x_{1}&+&a_{12}x_{2}&+&\dots &+&a_{1n}x_{n}&=b_{1},\\a_{21}x_{1}&+&a_{22}x_{2}&+&\dots &+&a_{2n}x_{n}&=b_{2},\\\vdots &&\vdots &&\ddots &&\vdots &\vdots \\a_{m1}x_{1}&+&a_{m2}x_{2}&+&\dots &+&a_{mn}x_{n}&=b_{m}.\end{matrix}}\end{cases}}.}$$

Wykorzystując macierze możemy to przedstawić następująco:

$${\begin{bmatrix}a_{11}&a_{12}&\dots &a_{1n}\\a_{21}&a_{22}&\dots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{m1}&a_{m2}&\dots &a_{mn}\end{bmatrix}}{\begin{bmatrix}x_{1}\\x_{2}\\\vdots \\x_{n}\end{bmatrix}}={\begin{bmatrix}b_{1}\\b_{2}\\\vdots \\b_{m}\end{bmatrix}}$$
I w zapisie skrótowym:
$$\mathbf {AX} =\mathbf {B}$$

Gdzie:  
$A = {\begin{bmatrix}a_{11}&a_{12}&\dots &a_{1n}\\a_{21}&a_{22}&\dots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{m1}&a_{m2}&\dots &a_{mn}\end{bmatrix}}$ - macierz współczynników

${\displaystyle \mathbf {B} =[b_{1},b_{2},\dots ,b_{m}]} $ - wektor wyrazów wolnych  
${\mathbf  X}=[x_{1},x_{2},\dots ,x_{n}] $ - wektor niewiadomych

Jeśli macierz układu $A$ jest macierzą kwadratową, to oznaczoność układu jest równoważna jej odwracalności, tzn.
$$\mathbf {AX} =\mathbf {B}$$
$${\displaystyle \mathbf {A} ^{-1}\mathbf {AX} =\mathbf {A} ^{-1}\mathbf {B}} $$
$$\mathbf {X} =\mathbf {A} ^{-1}\mathbf {B} .$$

#### Przykład:
Rozważmy układ równań:
$$\begin{cases}2x + 4y = 10 \\ x - y = -1 \end{cases}$$  
Rozwiązaniem jest para liczb:
$$\begin{cases}x = 1 \\ y = 2 \end{cases}$$  


### <a name='1'></a> Ćwiczenie 221
Wykorzystując bibliotekę numpy rozwiąż poniższy układ równań:

$$\begin{cases}5x - 3y = 21 \\ x - 2y = 7 \end{cases}$$

In [ ]:
# tutaj wpisz rozwiązanie
# Tworzymy macierz współczynników
A = np.array([[5, -3],
              [1, -2]])

# Tworzymy wektor wyrazów wolnych
B = np.array([21, 7])

# Rozwiązujemy układ równań
X = np.linalg.solve(A, B)

# Wyświetlamy rozwiązania
print("x = ", X[0])
print("y = ", X[1])

x =  3.0
y =  -2.0


Dokonaj sprawdzenia.

__Wskazówka:__ Użyj funkcji _np.allclose()_.

In [ ]:
# tutaj wpisz rozwiązanie
# Sprawdzenie: podstawiamy X do AX i porównujemy z B
is_correct = np.allclose(A @ X, B)
print("Czy rozwiązanie jest poprawne? :", is_correct)

Czy rozwiązanie jest poprawne? : True


### <a name='2'></a> Ćwiczenie 222
Wykorzystując bibliotekę numpy rozwiąż poniższy układ równań:

$$\begin{cases}x + y + z = 1 \\ 2x + y + 5z = 0 \\ x - y = z \end{cases}$$

In [ ]:
# tutaj wpisz rozwiązanie

# Macierz współczyński
A = np.array([
    [1, 1, 1],
    [2, 1, 5],
    [1, -1, -1]
])

# Wektor wyrazó wolnych
B = np.array([1, 0, 0])

# x + y + z = 1        ← tutaj **wyraz wolny to 1**
# 2x + y + 5z = 0      ← tutaj **wyraz wolny to 0**
# x - y = z            ← tę równość trzeba przekształcić -> x − y − z = 0


# Rozwiązanie układu
X = np.linalg.solve(A, B)

print("x = ", X[0])
print("y = ", X[1])
print("z = ", X[2])



x =  0.5
y =  0.875
z =  -0.375


Dokonaj sprawdzenia.

__Wskazówka:__ Użyj funkcji _np.allclose()_.

In [ ]:
# tutaj wpisz rozwiązanie
# Sprawdzenie poprawności rozwiązania
print("Czy rozwiązanie jest poprawne? :", np.allclose(A @ X, B))

Czy rozwiązanie jest poprawne? : True


### <a name='3'></a> Ćwiczenie 223
Załaduj popularny zbiór danych IRIS do zmiennej _data_raw_ wykorzystując bibliotekę _scikit-learn_.

In [ ]:
# tutaj wpisz rozwiązanie
from sklearn.datasets import load_iris

data_raw = load_iris()

Wyświetl opis zbioru IRIS

In [ ]:
# tutaj wpisz rozwiązanie
print(data_raw.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

Wyświetl nazwy zmiennych w zbiorze.

In [ ]:
# tutaj wpisz rozwiązanie
print(data_raw.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


Wyświetl nazwy klas w zbiorze.

In [ ]:
# tutaj wpisz rozwiązanie
print(data_raw.target_names)

### <a name='4'></a> Ćwiczenie 224
Do zmiennej _data_ przypisz dane (tablica numpy) zbioru IRIS.

In [ ]:
# tutaj wpisz rozwiązanie
data = data_raw.data

Do zmiennej _target_ przypisz wartości zmiennej docelowej ze zbioru IRIS.

In [ ]:
# tutaj wpisz rozwiązanie
target = data_raw.target

Zbadaj kształt zmiennej _data_ oraz _target_.

In [ ]:
# tutaj wpisz rozwiązanie
print("Shape of data:", data.shape)
print("Shape of target:", target.shape)


Shape of data: (150, 4)
Shape of target: (150,)


### <a name='5'></a> Ćwiczenie 225
Podziel dane na zbiór treningowy (_data_train_, _target_train_) i testowy (_data_test_, _target_test_). Rozmiar zbioru testowego to 30% próbek.

In [ ]:
# tutaj wpisz rozwiązanie
from sklearn.model_selection import train_test_split

data_train, data_test, target_train, target_test = train_test_split(data, target, test_size = 0.3, random_state = 20)

Wyświetl rozmiar zbiorów:
* _data_train_
* _target_train_
* _data_test_
* _target_test_

In [ ]:
# tutaj wpisz rozwiązanie
print("Rozmiar data_train:", len(data_train))
print("Rozmiar target_train:", len(target_train))
print("Rozmiar data_test:", len(data_test))
print("Rozmiar target_test:", len(target_test))


Rozmiar data_train: 105
Rozmiar target_train: 105
Rozmiar data_test: 45
Rozmiar target_test: 45


### <a name='6'></a> Ćwiczenie 226
Zbuduj model regresji logistcznej (ustaw domyślne wartości parametrów) wykorzystując bibliotekę scikit-learn oraz dane IRIS.

In [ ]:
# tutaj wpisz rozwiązanie
from sklearn.linear_model import LogisticRegression

# 1. Załaduj dane IRIS
iris = load_iris()
data = iris.data
target = iris.target

# 2. Podziel dane (test_size=0.3 oznacza 30% do testu)
data_train, data_test, target_train, target_test = train_test_split(data, target, test_size = 0.3, random_state = 20)

# 3. Zbuduj model regresji logistycznej z domyślnymi parametrami i wytrenuj go
model = LogisticRegression()

# Dopasuj dane treningowe do modelu
model.fit(data_train, target_train)

# 4. Wyświetl ocenę modelu (accuracy) na zbiorze testowym
accuracy = model.score(data_test, target_test)
print("Accuracy na zbiorze testowym: ", accuracy)

Accuracy na zbiorze testowym:  0.9333333333333333


**Opis kroków:**

**load_iris()** — wczytuje dane o kwiatach Iris (cechy + gatunki).

**train_test_split** — dzieli dane na część treningową i testową.

**LogisticRegression()** — tworzy model regresji logistycznej z domyślnymi parametrami.

**model.fit(...)** — trenuje model na zbiorze treningowym.

**model.score(...)** — zwraca accuracy (procent poprawnych klasyfikacji) na zbiorze testowym.

Dopasuj dane treningowe do modelu.

In [ ]:
# tutaj wpisz rozwiązanie
# Dopasuj dane treningowe do modelu
model.fit(data_train, target_train)

### <a name='7'></a> Ćwiczenie 227
Dokonaj oceny modelu na zbiorze treningowym.

In [ ]:
# tutaj wpisz rozwiązanie
accuracy_train = model.score(data_train, target_train)
print("Accuracy na zbiorze treiningowym: ", accuracy_train)

Accuracy na zbiorze treiningowym 0.9714285714285714


**Wyjaśnienie:**

* model.score(data_train, target_train) oblicza accuracy, czyli procent poprawnych przewidywań na treningowych danych.

* Im bliżej 1.0 (lub 100%), tym model lepiej „nauczył się” rozróżniać klasy na tych danych.

Dokonaj oceny modelu na zbiorze testowym.

In [ ]:
# tutaj wpisz rozwiązanie
accuracy_test = model.score(data_test, target_test)
print("Accuracy na zbiorze testowym: ", accuracy_test)

Accuracy na zbiorze testowym:  0.9333333333333333


### <a name='8'></a> Ćwiczenie 228
Dokonaj predykcji na danych testowych na podstawie modelu i przypisz je do zmiennej _target\_pred_.

In [ ]:
# tutaj wpisz rozwiązanie
# Dokonaj predykcji na danych testowych
target_pred = model.predict(data_test)


Wyświetl zmienną _target\_pred_.

In [ ]:
# tutaj wpisz rozwiązanie
# Wyświetl kilka pierwszych przewidywanych wartości
print("Predykcje danych testowych (pierwsze 10)", target_pred[ :10])

Predykcje danych testowych (pierwsze 10) [0 1 1 2 1 1 2 0 2 0]


Wyświetl zmienną _target\_test_.

In [ ]:
# tutaj wpisz rozwiązanie
print("Wartość przewidywaną (pierwsze 10)", target_test[ :10])

Wartość przewidywaną (pierwsze 10) [0 1 1 2 1 1 2 0 2 0]


### <a name='9'></a> Ćwiczenie 229
Wyznacz macierz pomyłek (macierz konfuzji)

In [ ]:
# tutaj wpisz rozwiązanie
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(target_test, target_pred)

print("Macierz pomyłek: ", cm)

Macierz pomyłek:  [[13  0  0]
 [ 0 18  0]
 [ 0  3 11]]


**Wyjaśnienie**:

* **confusion_matrix(y_true, y_pred)** przyjmuje najpierw rzeczywiste etykiety, potem przewidywane.

* Wynikowa macierz pokazuje, ile razy model dobrze lub źle zaklasyfikował próbki do poszczególnych klas.

Macierz pomyłek np: [
  
  [13  0  0]

 [ 0 18  0]

 [ 0  3 11]

 ]

##Jak czytać macierz pomyłek?
* **Wiersze**: Prawdziwe (rzeczywiste) klasy.
* **Kolumny**: Przewidywane przez model klasy.
* **Każdy element** [i, j] oznacza liczbę próbek, które należały do klasy i, a zostały zaklasyfikowane jako klasa j.

Przyjmijmy, że klasy są nazwane:
* **0** – Setosa
* **1** – Versicolor
* **2** – Virginica

### Rozbijmy macierz:

|                | Przewidziane: 0 | Przewidziane: 1 | Przewidziane: 2 |
|----------------|:---------------:|:---------------:|:---------------:|
| **Rzeczywista: 0** |      13        |        0        |        0        |
| **Rzeczywista: 1** |      0         |       18        |        0        |
| **Rzeczywista: 2** |      0         |        3        |       11        |

---

### Interpretacja:

1. **Klasa 0 (Setosa):**
   - **13** próbek klasy 0 zostało poprawnie sklasyfikowanych jako 0.
   - **0** zostało błędnie sklasyfikowanych jako 1 lub 2.

2. **Klasa 1 (Versicolor):**
   - **18** próbek klasy 1 zostało poprawnie sklasyfikowanych jako 1.
   - **0** zostało błędnie sklasyfikowanych jako 0 lub 2.

3. **Klasa 2 (Virginica):**
   - **11** próbek klasy 2 zostało poprawnie sklasyfikowanych jako 2.
   - **3** próbki klasy 2 zostały błędnie sklasyfikowane jako 1.


## Co to oznacza dla Twojego modelu?
* Model **bardzo dobrze rozpoznaje klasy 0 i 1** – nie myli ich w ogóle z innymi.

* **Klasa 2 (Virginica)** — model czasami myli ją z klasą 1 (Versicolor), zaklasyfikował 3 takie przypadki.

* **Nie ma żadnych pomyłek z klasą 0.**

## Wnioski:
* Wysoka skuteczność modelu.

* Jeśli zależy Ci na dokładnym rozpoznawaniu klasy 2 (Virginica), możesz pomyśleć o dalszej optymalizacji.

* Często Virginica jest trudniejsza do odróżnienia od Versicolor (stąd 3 pomyłki).

---


**Macierz pomyłek pozwala szczegółowo zobaczyć, gdzie model się myli i jakie klasy są najtrudniejsze do rozpoznania.**

W ten sposób można lepiej analizować skuteczność klasyfikatora niż tylko po samym accuracy!   

### <a name='10'></a> Ćwiczenie 230
Wyświetl raport klasyfikacji modelu.

In [ ]:
# tutaj wpisz rozwiązanie
from sklearn.metrics import classification_report

print(classification_report(target_test, target_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        13
           1       0.86      1.00      0.92        18
           2       1.00      0.79      0.88        14

    accuracy                           0.93        45
   macro avg       0.95      0.93      0.93        45
weighted avg       0.94      0.93      0.93        45



##Co zawiera raport?
* precision – precyzja (ile z przewidzianych jako dana klasa jest rzeczywiście tej klasy)

* recall – czułość (ile z rzeczywistych próbek danej klasy zostało poprawnie rozpoznanych)

* f1-score – średnia harmoniczna precyzji i recall

* support – liczba próbek danej klasy w zbiorze testowym

---

#### **1. accuracy**

- **accuracy** (dokładność, trafność) – to po prostu odsetek poprawnych klasyfikacji:
  
  \[
  \text{accuracy} = \frac{\text{liczba poprawnie sklasyfikowanych próbek}}{\text{wszystkie próbki}}
  \]
- W powyższym przykładzie:  
  `0.93` oznacza, że **93%** próbek zostało poprawnie sklasyfikowanych.
- Liczba `45` oznacza, że w zbiorze testowym było **45 próbek**.

---

#### **2. macro avg (średnia makro)**

- Najpierw liczona jest precision, recall i f1-score osobno dla każdej klasy,  
  **potem wyciągana jest z nich średnia arytmetyczna** (po klasach).
- Każda klasa jest **ważona tak samo**, niezależnie od liczby próbek danej klasy.
- Dobra miara, jeśli zależy Ci, by wszystkie klasy traktować równo (ważne przy niezbalansowanych danych).

---

#### **3. weighted avg (średnia ważona)**

- Najpierw liczysz precision, recall i f1-score dla każdej klasy,
- Następnie każda wartość jest **ważona liczbą próbek w tej klasie (support)**.
- Klasy z większą liczbą próbek mają większy wpływ na końcowy wynik.

---

#### **Porównanie macro avg vs weighted avg:**

- Jeśli Twoje klasy są niezbalansowane (jedna klasa dominuje), **weighted avg** lepiej oddaje ogólną jakość predykcji.
- **macro avg** jest czuła na to, czy któraś klasa jest szczególnie źle klasyfikowana, nawet jeśli takich próbek jest mało.

---

| Metryka       | Opis                                                                          |
|---------------|-------------------------------------------------------------------------------|
| accuracy      | Ogólna trafność klasyfikacji (proporcja poprawnych predykcji dla wszystkich). |
| macro avg     | Średnia arytmetyczna metryk dla każdej klasy (wszystkie klasy równo ważone).  |
| weighted avg  | Średnia ważona metryk, ważona liczbą próbek w każdej klasie (support).        |